src/main/webapp/app/exam/participate/exercises/exam-exercise-update-highlighter/exam-exercise-update-highlighter.component.ts

```diff
@@ -0,0 +1,118 @@
+        } else {
+            this.exercise.problemStatement = this.updatedProblemStatementWithHighlightedDifferences;
+        }
+        this.showHighlightedDifferences = !this.showHighlightedDifferences;
+        this.problemStatementUpdateEvent.emit(this.exercise.problemStatement);
+    }
+
+    /**
+     * Updates the problem statement of the provided exercises based on its id.
+     * Also calls the method to highlight the differences between the old and new problem statement.
+     * @param exerciseId is the id of the exercise which problem statement should be updated.
+     * @param updatedProblemStatement is the new problem statement that should replace the old one.
+     */
+    updateExerciseProblemStatementById(exerciseId: number, updatedProblemStatement: string) {
+        if (updatedProblemStatement != undefined && exerciseId === this.exercise.id) {
+            this.updatedProblemStatement = updatedProblemStatement;
+            this.exercise.problemStatement = this.highlightProblemStatementDifferences();
+        }
+        this.problemStatementUpdateEvent.emit(this.exercise.problemStatement);
+    }
+
+    /**
+     * Computes the difference between the old and new (updated) problem statement and displays this difference.
+     */
+    highlightProblemStatementDifferences() {
+        if (!this.updatedProblemStatement) {
+            return;
+        }
+
+        this.showHighlightedDifferences = true;
+
+        // creates the diffMatchPatch library object to be able to modify strings
+        const dmp = new DiffMatchPatch();
+        let outdatedProblemStatement: string;
+
+        // checks if first update i.e. no highlight
+        if (!this.previousProblemStatementUpdate) {
+            outdatedProblemStatement = this.exercise.problemStatement!;
+            // else use previousProblemStatementUpdate as new outdatedProblemStatement to avoid inserted HTML elements
+        } else {
+            outdatedProblemStatement = this.previousProblemStatementUpdate;
+        }
+
+        this.previousProblemStatementUpdate = this.updatedProblemStatement;
+
+        // finds the initial difference then cleans the text with added html & css elements
+        const diff = dmp.diff_main(outdatedProblemStatement!, this.updatedProblemStatement);
+        dmp.diff_cleanupEfficiency(diff);
+        // remove ¶; (= &para;) symbols
+
+        this.updatedProblemStatementWithHighlightedDifferences = this.diffPrettyHtml(diff);
+        return this.updatedProblemStatementWithHighlightedDifferences;
+    }
+
+    /**
+     * Convert a diff array into a pretty HTML report.
+     * Modified diff_prettHtml() method from DiffMatchPatch
+     * Keeps markdown styling intact (not like the original method)
+     * @param diffs Array of diff tuples. (from DiffMatchPatch)
+     * @return the HTML representation as string with markdown intact.
+     */
+    diffPrettyHtml = function (diffs: Diff[]): string {
```

Comment:
You could create an interface fo the Diff, e.g.
```
interface Diff {
     operation: string // or maybe even better an enum type for the three cases,
     text: string
}
````
Then you would not need the comments when working with this interfaced type Diff, but could rather do:
`diffs[x|.operation` and `diffs[x].text`

Reply:
but see my comment below to refactor the method in general and avoid accessing objects in the array via index.

Reply:
as you want to iterate over the item in an array and keep track of the index you could also do:
```

diffs.forEach((diff, index) => {
     switch(diff.operation) {
           case DiffOperation.DIFF_INSERT:
             ...
     }
}
``` 

You would not need to access the elements of the array by index, and you would not need to assign `op` and `text`.
So overall, I would suggest something like:
```
enum DiffOperation {
    DIFF_INSERT = 'DIFF_INSERT',
    DIFF_DELETE = 'DIFF_DELETE',
    DIFF_EQUAL = 'DIFF_EQUAL,'
}

interface Diff {
    operation: DiffOperation,
    text: string,
}
let diff1: Diff = {
    operation: DiffOperation.DIFF_INSERT,
    text: 'Hi',
}

let diff2: Diff = {
    operation: DiffOperation.DIFF_DELETE,
    text: 'Ho',
}

let diffs = [diff1, diff2]

let diffPrettyHtml = function (diffs: Diff[]): string {
    let html: string[] = [];
    diffs.forEach((diff: Diff, index: number) => {
        switch(diff.operation) {
            case DiffOperation.DIFF_INSERT:
                html[index] = '<ins style="background:#e6ffe6;">' + diff.text + '</ins>';
                break;
            case DiffOperation.DIFF_DELETE:
                html[index] = '<del style="background:#ffe6e6;">' + diff.text + '</del>';
                break;
            case DiffOperation.DIFF_EQUAL:
                html[index] = diff.text;
                break;
            }
        }
    )
    return html.join('');
}
```
I added the enum and some examples to try it out myself, I left it in the example but I know that the enum already exist and you import it :)

Reply:
Good idea.
I will change the for loop to a for each loop but I want to stick to the current types and not introduce new interfaces even if it would be easier but I want to keep the original implementation of this library method and stick to it as close as possible and reuse the already provided/imported types.
https://github.com/google/diff-match-patch/blob/master/javascript/diff_match_patch_uncompressed.js
https://www.npmjs.com/package/diff-match-patch
![image](https://user-images.githubusercontent.com/65814168/132847266-5c998b57-243b-4e7d-b4bf-94fb4db3e4a1.png)